File: fetch_knowledgeboat_videos.js
/**
 * Fetch all playlists and videos from the YouTube channel "Knowledge Boat"
 * and write a grouped JSON file consumable by the website's videos page.
 *
 * Usage:
 *   node fetch_knowledgeboat_videos.js --out public/videos.json --channel "Knowledge Boat"
 *   # or set YT_API_KEY via .env or environment
 *
 * Notes:
 *  - Uses only Node built-ins (https/url/fs) + a tiny .env reader below (no deps).
 *  - Groups by playlist; each playlist title is used as category header.
 */

const fs = require('fs');
const https = require('https');
const { URL } = require('url');
const path = require('path');

const YOUTUBE_API_BASE = 'https://www.googleapis.com/youtube/v3';

function loadDotEnv(dotenvPath = '.env') {
  try {
    if (!fs.existsSync(dotenvPath)) return;
    const txt = fs.readFileSync(dotenvPath, 'utf8');
    for (const rawLine of txt.split(/\r?\n/)) {
      const line = rawLine.trim();
      if (!line || line.startsWith('#') || !line.includes('=')) continue;
      const idx = line.indexOf('=');
      const key = line.slice(0, idx).trim();
      let val = line.slice(idx + 1).trim();
      if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
        val = val.slice(1, -1);
      }
      if (key && !(key in process.env)) process.env[key] = val;
    }
  } catch {}
}

loadDotEnv();

function httpGetJson(endpoint, params) {
  const url = new URL(`${YOUTUBE_API_BASE}/${endpoint}`);
  Object.entries(params).forEach(([k, v]) => {
    if (v !== undefined && v !== null && v !== '') url.searchParams.set(k, String(v));
  });
  return new Promise((resolve, reject) => {
    https.get(url, (res) => {
      const status = res.statusCode || 0;
      const chunks = [];
      res.on('data', (d) => chunks.push(d));
      res.on('end', () => {
        const body = Buffer.concat(chunks).toString('utf8');
        if (status < 200 || status >= 300) {
          let details = '';
          try { details = JSON.parse(body)?.error?.message || body; } catch { details = body; }
          const err = new Error(`HTTP ${status} for ${url} -> ${details}`);
          err.statusCode = status;
          err.responseBody = body;
          return reject(err);
        }
        try {
          resolve(JSON.parse(body));
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

async function findChannelId(apiKey, channelName) {
  const search = await httpGetJson('search', {
    part: 'snippet',
    q: channelName,
    type: 'channel',
    maxResults: 5,
    key: apiKey,
  });
  const items = search.items || [];
  if (!items.length) throw new Error(`No channel results for query: ${channelName}`);
  const normalized = channelName.trim().toLowerCase();
  const exact = items.find((it) => (it.snippet?.title || '').trim().toLowerCase() === normalized);
  const target = exact || items[0];
  return target.snippet.channelId;
}

async function listPlaylists(apiKey, channelId) {
  const result = [];
  let pageToken = undefined;
  do {
    const payload = await httpGetJson('playlists', {
      part: 'snippet,contentDetails',
      channelId,
      maxResults: 50,
      pageToken,
      key: apiKey,
    });
    for (const it of payload.items || []) {
      result.push({
        id: it.id,
        title: it.snippet?.title || '',
        description: it.snippet?.description || '',
        itemCount: it.contentDetails?.itemCount,
      });
    }
    pageToken = payload.nextPageToken;
  } while (pageToken);
  return result;
}

async function listVideosInPlaylist(apiKey, playlistId) {
  const videos = [];
  let pageToken = undefined;
  do {
    const payload = await httpGetJson('playlistItems', {
      part: 'snippet,contentDetails',
      playlistId,
      maxResults: 50,
      pageToken,
      key: apiKey,
    });
    for (const it of payload.items || []) {
      const snippet = it.snippet || {};
      const content = it.contentDetails || {};
      const videoId = content.videoId;
      if (!videoId) continue;
      const thumbs = snippet.thumbnails || {};
      const thumbUrl =
        thumbs.maxres?.url ||
        thumbs.standard?.url ||
        thumbs.high?.url ||
        thumbs.medium?.url ||
        thumbs.default?.url ||
        `https://i.ytimg.com/vi/${videoId}/hqdefault.jpg`;
      videos.push({
        id: videoId,
        title: snippet.title || '',
        description: snippet.description || '',
        publishedAt: snippet.publishedAt,
        thumbnail: thumbUrl,
        position: snippet.position,
      });
    }
    pageToken = payload.nextPageToken;
  } while (pageToken);
  return videos;
}

function fetchBody(u) {
  return new Promise((resolve, reject) => {
    https.get(u, (res) => {
      const chunks = [];
      res.on('data', (d) => chunks.push(d));
      res.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    }).on('error', reject);
  });
}

function extractChannelIdFromHtml(html) {
  const m1 = html.match(/"channelId":"(UC[0-9A-Za-z_-]{20,})"/);
  if (m1 && m1[1]) return m1[1];
  const m2 = html.match(/"externalId":"(UC[0-9A-Za-z_-]{20,})"/);
  if (m2 && m2[1]) return m2[1];
  const m3 = html.match(/\/[cC]hannel\/(UC[0-9A-Za-z_-]{20,})/);
  if (m3 && m3[1]) return m3[1];
  return '';
}

async function fetchChannelIdFromHandle(handle) {
  const handleClean = handle.startsWith('@') ? handle : `@${handle}`;
  const tryUrls = [
    `https://www.youtube.com/${handleClean}`,
    `https://www.youtube.com/${handleClean}/about`,
  ];
  for (const u of tryUrls) {
    const body = await fetchBody(u);
    const id = extractChannelIdFromHtml(body);
    if (id) return id;
  }
  throw new Error(`Could not resolve channelId from handle ${handleClean}`);
}

async function fetchChannelIdFromUrl(url) {
  const body = await fetchBody(url);
  const id = extractChannelIdFromHtml(body);
  if (id) return id;
  throw new Error(`Could not resolve channelId from url ${url}`);
}

async function buildChannelCatalog(apiKey, channelName, opts = {}) {
  let channelId = opts.channelId || '';
  // If user passed a UC* id via --channel, use it directly
  if (!channelId && /^UC[0-9A-Za-z_-]{20,}$/.test(channelName)) channelId = channelName;
  // If a handle was provided, resolve it without hitting the API search
  if (!channelId && opts.handle) {
    channelId = await fetchChannelIdFromHandle(opts.handle);
  }
  // If a full YouTube URL was provided, resolve from page source
  if (!channelId && opts.url) {
    channelId = await fetchChannelIdFromUrl(opts.url);
  }
  if (!channelId) channelId = await findChannelId(apiKey, channelName);
  const playlists = await listPlaylists(apiKey, channelId);
  const categories = [];
  for (const pl of playlists) {
    const vids = await listVideosInPlaylist(apiKey, pl.id);
    if (!vids.length) continue;
    const vidsSorted = [...vids].sort((a, b) => {
      const ap = a.publishedAt || '';
      const bp = b.publishedAt || '';
      if (ap === bp) return (b.position || 0) - (a.position || 0);
      return ap < bp ? 1 : -1;
    });
    categories.push({ id: pl.id, title: pl.title, videos: vidsSorted });
    await new Promise((r) => setTimeout(r, 150));
  }
  categories.sort((a, b) => a.title.toLowerCase().localeCompare(b.title.toLowerCase()));
  return {
    channel: { name: channelName, id: channelId },
    categories,
    generatedAt: new Date().toISOString().replace(/\..+/, 'Z'),
  };
}

async function main() {
  const argv = process.argv.slice(2);
  const outIdx = argv.indexOf('--out');
  const outPath = outIdx !== -1 ? argv[outIdx + 1] : path.join('public', 'videos.json');
  const chIdx = argv.indexOf('--channel');
  const channelName = chIdx !== -1 ? argv[chIdx + 1] : 'Knowledge Boat';
  const idIdx = argv.indexOf('--channelId');
  const channelId = idIdx !== -1 ? argv[idIdx + 1] : '';
  const handleIdx = argv.indexOf('--handle');
  const handle = handleIdx !== -1 ? argv[handleIdx + 1] : '';
  const urlIdx = argv.indexOf('--url');
  const url = urlIdx !== -1 ? argv[urlIdx + 1] : '';
  const apiKey = process.env.YT_API_KEY || '';
  if (!apiKey) {
    console.error('Missing API key. Set YT_API_KEY in .env or environment.');
    process.exit(2);
  }

  try {
    const catalog = await buildChannelCatalog(apiKey, channelName, { channelId, handle, url });
    const dir = path.dirname(outPath);
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(outPath, JSON.stringify(catalog, null, 2), 'utf8');
    const total = catalog.categories.reduce((acc, c) => acc + (c.videos?.length || 0), 0);
    console.log(`Wrote ${outPath} with ${total} videos across ${catalog.categories.length} categories.`);
  } catch (e) {
    console.error('Error while fetching data:', e?.message || e);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

File: fetch_knowledgeboat_videos.py
"""
Fetch all playlists and videos from the YouTube channel "Knowledge Boat" and
write a grouped JSON file consumable by the website's videos page.

Usage (API key default is embedded but can be overridden):
  python fetch_knowledgeboat_videos.py \
      --api-key YOUR_KEY \
      --channel-name "Knowledge Boat" \
      --out "public/videos.json"

Notes:
  - Uses YouTube Data API v3.
  - Groups videos by playlist; each playlist title is used as a category header.
  - Only relies on Python standard library (urllib) — no external deps.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.parse
import urllib.request
from typing import Dict, List, Optional, Any


# Load API key from environment; recommend setting in a local .env file
DEFAULT_API_KEY = os.getenv("YT_API_KEY", "")

YOUTUBE_API_BASE = "https://www.googleapis.com/youtube/v3"


def http_get_json(endpoint: str, params: Dict[str, Any]) -> Dict[str, Any]:
    query = urllib.parse.urlencode({k: v for k, v in params.items() if v is not None})
    url = f"{YOUTUBE_API_BASE}/{endpoint}?{query}"
    with urllib.request.urlopen(url) as resp:
        data = resp.read()
    return json.loads(data.decode("utf-8"))


def find_channel_id(api_key: str, channel_name: str) -> str:
    # First try search.list to find the channel by query
    search = http_get_json(
        "search",
        {
            "part": "snippet",
            "q": channel_name,
            "type": "channel",
            "maxResults": 5,
            "key": api_key,
        },
    )
    items = search.get("items", [])
    if not items:
        raise RuntimeError(f"No channel results for query: {channel_name}")

    # Prefer exact title match (case-insensitive), else take first result
    normalized = channel_name.strip().lower()
    exact = next((it for it in items if it["snippet"]["title"].strip().lower() == normalized), None)
    target = exact or items[0]
    return target["snippet"]["channelId"]


def list_playlists(api_key: str, channel_id: str) -> List[Dict[str, Any]]:
    playlists: List[Dict[str, Any]] = []
    page_token: Optional[str] = None
    while True:
        payload = http_get_json(
            "playlists",
            {
                "part": "snippet,contentDetails",
                "channelId": channel_id,
                "maxResults": 50,
                "pageToken": page_token,
                "key": api_key,
            },
        )
        for it in payload.get("items", []):
            playlists.append(
                {
                    "id": it["id"],
                    "title": it["snippet"]["title"],
                    "description": it["snippet"].get("description", ""),
                    "itemCount": it.get("contentDetails", {}).get("itemCount"),
                }
            )
        page_token = payload.get("nextPageToken")
        if not page_token:
            break
        time.sleep(0.1)
    return playlists


def list_videos_in_playlist(api_key: str, playlist_id: str) -> List[Dict[str, Any]]:
    videos: List[Dict[str, Any]] = []
    page_token: Optional[str] = None
    while True:
        payload = http_get_json(
            "playlistItems",
            {
                "part": "snippet,contentDetails",
                "playlistId": playlist_id,
                "maxResults": 50,
                "pageToken": page_token,
                "key": api_key,
            },
        )
        for it in payload.get("items", []):
            snippet = it.get("snippet", {})
            content = it.get("contentDetails", {})
            video_id = content.get("videoId")
            if not video_id:
                continue
            # Prefer high-res thumbnail if present
            thumbs = snippet.get("thumbnails", {})
            thumb_url = (
                (thumbs.get("maxres") or {}).get("url")
                or (thumbs.get("standard") or {}).get("url")
                or (thumbs.get("high") or {}).get("url")
                or (thumbs.get("medium") or {}).get("url")
                or (thumbs.get("default") or {}).get("url")
                or f"https://i.ytimg.com/vi/{video_id}/hqdefault.jpg"
            )
            videos.append(
                {
                    "id": video_id,
                    "title": snippet.get("title", ""),
                    "description": snippet.get("description", ""),
                    "publishedAt": snippet.get("publishedAt"),
                    "thumbnail": thumb_url,
                    "position": snippet.get("position"),
                }
            )
        page_token = payload.get("nextPageToken")
        if not page_token:
            break
        time.sleep(0.1)
    return videos


def build_channel_catalog(api_key: str, channel_name: str) -> Dict[str, Any]:
    channel_id = find_channel_id(api_key, channel_name)
    playlists = list_playlists(api_key, channel_id)

    categories = []
    for pl in playlists:
        vids = list_videos_in_playlist(api_key, pl["id"])
        if not vids:
            continue
        # Newest first within playlist (publishedAt may be None for some items, fallback by position)
        vids_sorted = sorted(
            vids,
            key=lambda v: (v.get("publishedAt") or "", -(v.get("position") or 0)),
            reverse=True,
        )
        categories.append(
            {
                "id": pl["id"],
                "title": pl["title"],
                "videos": vids_sorted,
            }
        )
        # Be gentle with API quotas
        time.sleep(0.15)

    # Sort categories alphabetically by title for consistent UI (can be adjusted later)
    categories = sorted(categories, key=lambda c: c["title"].lower())

    return {
        "channel": {
            "name": channel_name,
            "id": channel_id,
        },
        "categories": categories,
        "generatedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    }


def main(argv: Optional[List[str]] = None) -> int:
    # Load environment variables from a local .env if present (no external deps)
    def load_env_from_dotenv(path: str = ".env") -> None:
        if not os.path.exists(path):
            return
        try:
            with open(path, "r", encoding="utf-8") as f:
                for raw in f:
                    line = raw.strip()
                    if not line or line.startswith("#") or "=" not in line:
                        continue
                    key, value = line.split("=", 1)
                    key = key.strip()
                    value = value.strip().strip('"').strip("'")
                    # Do not override if already set in environment
                    if key and key not in os.environ:
                        os.environ[key] = value
        except Exception:
            # Non-fatal: ignore malformed .env
            pass

    load_env_from_dotenv()
    parser = argparse.ArgumentParser(description="Fetch YouTube channel playlists/videos and write JSON")
    parser.add_argument("--api-key", dest="api_key", default=os.getenv("YT_API_KEY", DEFAULT_API_KEY))
    parser.add_argument("--channel-name", dest="channel_name", default="Knowledge Boat")
    parser.add_argument("--out", dest="out_path", default=os.path.join("public", "videos.json"))
    args = parser.parse_args(argv)

    if not args.api_key:
        print("Missing API key. Provide --api-key or set YT_API_KEY.", file=sys.stderr)
        return 2

    try:
        catalog = build_channel_catalog(args.api_key, args.channel_name)
    except Exception as exc:
        print(f"Error while fetching data: {exc}", file=sys.stderr)
        return 1

    os.makedirs(os.path.dirname(args.out_path), exist_ok=True)
    with open(args.out_path, "w", encoding="utf-8") as f:
        json.dump(catalog, f, ensure_ascii=False, indent=2)
    print(f"Wrote {args.out_path} with {sum(len(c['videos']) for c in catalog['categories'])} videos across {len(catalog['categories'])} categories.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())


